lncrawl/sources/asadatrans.py

```diff
@@ -0,0 +1,102 @@
+
+    def search_novel(self, query):
+        query = query.lower().replace(' ', '+')
+        soup = self.get_soup(search_url % query)
+
+        results = []
+        for tab in soup.select('.c-tabs-item__content'):
+            a = tab.select_one('.post-title h3 a')
+            latest = tab.select_one('.latest-chap .chapter a').text
+            votes = tab.select_one('.rating .total_votes').text
+            results.append({
+                'title': a.text.strip(),
+                'url': self.absolute_url(a['href']),
+                'info': '%s | Rating: %s' % (latest, votes),
+            })
+        # end for
+
+        return results
+    # end def
+
+    def read_novel_info(self):
+        '''Get novel title, autor, cover etc'''
+        logger.debug('Visiting %s', self.novel_url)
+        soup = self.get_soup(self.novel_url)
+
+        possible_title = soup.select_one('.post-title h1')
+        for span in possible_title.select('span'):
+            span.extract()
+        # end for
+        self.novel_title = possible_title.text.strip()
+        logger.info('Novel title: %s', self.novel_title)
+
+        # NOTE: Site doesn't have book covers.
+        # self.novel_cover = self.absolute_url(
+        #     soup.select_one('.summary_image a img')['src'])
+        # logger.info('Novel cover: %s', self.novel_cover)
+
+        self.novel_author = ' '.join([
+            a.text.strip()
+            for a in soup.select('.author-content a[href*="manga-author"]')
+        ])
+        logger.info('%s', self.novel_author)
+
+        volumes = set()
+        chapters = soup.select('ul.main li.wp-manga-chapter a')
+        for a in reversed(chapters):
+            chap_id = len(self.chapters) + 1
+            vol_id = (chap_id - 1) // 100 + 1
+            volumes.add(vol_id)
+            self.chapters.append({
+                'id': chap_id,
+                'volume': vol_id,
+                'url':  self.absolute_url(a['href']),
+                'title': a.text.strip() or ('Chapter %d' % chap_id),
+            })
+        # end for
+
+        self.volumes = [{'id': x} for x in volumes]
+    # end def
+
+    def download_chapter_body(self, chapter):
+        '''Download body of a single chapter and return as clean html format.'''
+        logger.info('Downloading %s', chapter['url'])
+        soup = self.get_soup(chapter['url'])
+
+        contents = soup.select_one('div.text-left')
+        for bad in contents.select('h3, .code-block, script, .adsbygoogle, .sharedaddy'):
+            bad.decompose()
+
+        # remove bad text
+        self.blacklist_patterns = [
+            r'^Translator:',
+            r'^Qii',
+            r'^Editor:',
+            r'^Maralynx',
+            r'^Translator and Editor Notes:',
+            r'^Support this novel on',
+            r'^NU',
+            r'^by submitting reviews and ratings or by adding it to your reading list.',
+        ]
+
+        for discord in contents.select('p'):
```

Comment:
Can't you add `p` element (preferably with some attribute value) into the `.select()` on line 79 and do a single loop? You have 3 loops here which seems a bit yucky and very expensive because of how many `p` elements each chapter has.
edit:
looking at that site, they seem to have the first `p` child with that content:
```html
<p class="has-text-color has-primary-color">Join our <a href="https://discord.gg/TFQGdmy">discord</a> to get latest updates about the novel</p>
```
maybe just get the first child, check if it has a discord link and remove it. Or use those class selectors?